### ISSUE TITLE
Liquid haskell is unable to find equality in eta reduction

### ISSUE THREAD

**Issue by mckeankylej** [ucsd-progsys/liquidhaskell#791]
When liquid type checking the following program
```haskell
{-@ measure mvlen :: forall a. (MVector s a) -> Int @-}
{-@ invariant {mv:MVector s a | 0 <= mvlen mv } @-}
{-@ assume M.new :: PrimMonad m => n:Nat -> m ({vs:MVector (PrimState m) a | mvlen vs == n }) @-}
{-@ minimal :: v:Vector a -> Vector a @-}
minimal :: Vector a -> Vector a
minimal v = create (M.new (V.length v) >>= \target -> go target)
    {-@ go :: in:{in:MVector (PrimState m) a | mvlen in == vlen v } -> m (MVector (PrimState m) a) @-}
  where go :: (PrimMonad m) => MVector (PrimState m) a ->
                               m (MVector (PrimState m) a)
        go mv = pure mv

```
I get the following error
```sh
**** RESULT: UNSAFE ************************************************************

 Unique.hs:138:55-63: Error: Liquid Type Mismatch
 
 138 | minimal v = create (M.new (V.length v) >>= \target -> go target)
                                                             ^^^^^^^^^
 
   Inferred type
     VV : {VV : (Data.Vector.Mutable.MVector (Control.Monad.Primitive.PrimState (GHC.ST.ST a)) b) | VV == ?a
                                                                                                    && 0 <= mvlen VV}
  
   not a subtype of Required type
     VV : {VV : (Data.Vector.Mutable.MVector (Control.Monad.Primitive.PrimState (GHC.ST.ST a)) b) | mvlen VV == vlen v}
  
   In Context
     ?a := {?a : (Data.Vector.Mutable.MVector (Control.Monad.Primitive.PrimState (GHC.ST.ST a)) b) | ?a ~~ target
                                                                                                     && 0 <= mvlen ?a}

     v := {v : (Data.Vector.Vector a) | 0 <= vlen v}

     target := {target : (Data.Vector.Mutable.MVector a b) | 0 <= mvlen target}

```
But if I eta reduce the program to look like this
```haskell
{-@ measure mvlen :: forall a. (MVector s a) -> Int @-}
{-@ invariant {mv:MVector s a | 0 <= mvlen mv } @-}
{-@ assume M.new :: PrimMonad m => n:Nat -> m ({vs:MVector (PrimState m) a | mvlen vs == n }) @-}
{-@ minimal :: v:Vector a -> Vector a @-}
minimal :: Vector a -> Vector a
minimal v = create (M.new (V.length v) >>= go)
    {-@ go :: in:{in:MVector (PrimState m) a | mvlen in == vlen v } -> m (MVector (PrimState m) a) @-}
  where go :: (PrimMonad m) => MVector (PrimState m) a ->
                               m (MVector (PrimState m) a)
        go mv = pure mv
```
The code liquid type checks!

**Comment by ranjitjhala** [comment]
Hmm, thanks for pointing this out @mckeankylej! 

1. Can you add the `import`s to the above so I can try to reproduce?
2. Are you running this on `master` or `develop` ? 

Off the top of my head, I wonder if the reason is 
similar to whats going on here 

https://github.com/ucsd-progsys/liquidhaskell/issues/675

But I'd like to run it to be sure! 

Thanks!

**Comment by ranjitjhala** [comment]
Also: as in #675 -- can you run with `--scrape-imports` to see if that makes a difference? Thanks!

**Comment by mckeankylej** [comment]
I am running whatever is the latest version on hackage I can't seem to build from source due to a nixos specific bug. I think I am going to open an issue for that too but I want to see if I can solve that first. 
Here are the relevant imports
```haskell
import Prelude hiding (null, head, tail, notElem, empty)

import Data.Vector hiding (singleton, empty)
import qualified Data.Vector.Mutable as M
import Control.Monad.Primitive
import qualified Data.Vector as V
```
Running with scrape-imports errored out with the above error.

**Comment by ranjitjhala** [comment]
Ok, I can reproduce this now, thanks!

There are *two* separate bugs here (tests/todo/T791.hs) and (tests/todo/T791a.hs) 

1. Why does the verification fail in the eta-expanded version? Am puzzled by where the refinement is getting lost,

2. More importantly, the verification *should fail but doesn't* in the unexpanded version, in fact in T791a.hs, I have further simplified it to the below which is obviously wrong but returns safe.

@nikivazou can you look into the latter at least ASAP? It is very worrying! 

```haskell
import Prelude hiding (null, head, tail, notElem, empty)

import Data.Vector hiding (singleton, empty)
import qualified Data.Vector.Mutable as M
import Control.Monad.Primitive
import qualified Data.Vector as V

{-@ measure mvlen :: forall a. (MVector s a) -> Int @-}

{-@ minimal1 :: v:Vector a -> Vector a @-}
minimal1 :: Vector a -> Vector a
minimal1 mickey = create (M.new 0 >>= \zong -> go zong)
    {-@ go :: {v:MVector (PrimState m) a | false } -> m (MVector (PrimState m) a) @-}
  where go :: (PrimMonad m) => MVector (PrimState m) a -> m (MVector (PrimState m) a)
        go mv = pure mv
```

**Comment by ranjitjhala** [comment]
@nikivazou  in the above (unsound) case there is no subtyping constraint even created of the form `G |- ... <: false` so it must be a constraint generation bug.

**Comment by ranjitjhala** [comment]
I think the problem here -- in *both cases* -- is our limited support for `Cast` expressions. Looking at the core for the above, now further simplified to:

```haskell
import Prelude hiding (null, head, tail, notElem, empty)

import Data.Vector hiding (singleton, empty)
import qualified Data.Vector.Mutable as M
import Control.Monad.Primitive
import qualified Data.Vector as V

minimal1 mickey = create (M.new 0 >>= go)

{-@ go :: {v:MVector (PrimState m) a | false } -> m (MVector (PrimState m) a) @-}
go :: (PrimMonad m) => MVector (PrimState m) a -> m (MVector (PrimState m) a)
go mv = pure mv
```

I see that the `e >>= go` term is converted to the `CoreExpr`

```
  let {
             lq_anf$##1677733835
               :: Data.Vector.Mutable.MVector
                    (Control.Monad.Primitive.PrimState (GHC.ST.ST s)) a
                  -> GHC.ST.ST
                       s
                       (Data.Vector.Mutable.MVector
                          (Control.Monad.Primitive.PrimState (GHC.ST.ST s)) a)
             [LclId, Str=DmdType]
             lq_anf$##1677733835 =
               Main.go @ (GHC.ST.ST s) @ a $dPrimMonad_a33S } in
           let {
             lq_anf$##1677733836
               :: Data.Vector.Mutable.MVector s a
                  -> GHC.ST.ST s (Data.Vector.Mutable.MVector s a)
             [LclId, Str=DmdType]
             lq_anf$##1677733836 =
               lq_anf$##1677733835
               `cast` ((Data.Vector.Mutable.MVector
                          (Control.Monad.Primitive.TFCo:R:PrimStateST[0] <s>_N) <a>_R)_R
                       -> (GHC.ST.ST
                             <s>_N
                             (Data.Vector.Mutable.MVector
                                (Control.Monad.Primitive.TFCo:R:PrimStateST[0] <s>_N) <a>_R)_R)_R
                       :: (Data.Vector.Mutable.MVector
                             (Control.Monad.Primitive.PrimState (GHC.ST.ST s)) a
                           -> GHC.ST.ST
                                s
                                (Data.Vector.Mutable.MVector
                                   (Control.Monad.Primitive.PrimState (GHC.ST.ST s)) a))
                          ~R# (Data.Vector.Mutable.MVector s a
                               -> GHC.ST.ST s (Data.Vector.Mutable.MVector s a))) } in
```

The strange thing there is the function cast, which in `castTy` 
used in `Generate.hs` will essentially strip out the refinements (it just preserves the 
thereby stripping out the precondition, thereby causing LH to 
say SAFE.

In the eta-expanded case `e >>= (\x -> go x)` the `cast` is applied to the `e` 
and again, we strip out all the refinements (losing the `mvlen` refinement from 
the `new`) which is why LH says unsafe.

In short: our current limited handling of `Cast` is both unsound and imprecise, 
and so we need to properly figure out the semantics of these expressions and 
fix this problem properly.

[Thanks @mckeankylej for pointing this out -- unfortunately it 
may take a while to fix, and I'm not sure what a good work-around 
would be, as I don't really understand why the `Cast` is getting 
inserted... perhaps because the class constraint is not at the top-level?
There were various examples like this in our regression tests in 
`benchmarks/vector-algorithms` but I think all have the class 
constraints at the top-level...]

**Comment by ranjitjhala** [comment]
@mckeankylej for the moment, one workaround is to not use the `PrimState` and `PrimMonad` constraints, but just use `ST` so, the following version of your code -- where I have specialized all the `m (MVector (PrimState m) a)` to `ST s (MVector s a)` is accepted as SAFE by LH both with 
and without eta-expanding the `go` ...

```haskell
{-@ measure mvlen :: forall a. (MVector s a) -> Int @-}
{-@ invariant {mv:MVector s a | 0 <= mvlen mv } @-}

{-@ assume mutNew :: n:Nat -> ST s ({v:MVector s a | mvlen v == n}) @-}
mutNew :: Int -> ST s (MVector s a)
mutNew = M.new

{-@ minimal :: v:Vector a -> Vector a @-}
minimal :: Vector a -> Vector a
minimal v = create (mutNew {- M.new -} (V.length v) >>= (\x -> go x))
  where
    {-@ go :: {mv:MVector s a | mvlen mv == vlen v} -> ST s (MVector s a) @-}
    go :: MVector s a -> ST s (MVector s a)
    go mv = pure mv
```

**Comment by mckeankylej** [comment]
The above workaround works! Its sad that you have to make a new version of `new` with a less specific type though.

**Comment by ranjitjhala** [comment]
Yes it's unfortunate indeed -- but we should be able to handle it once we
figure out what the semantics of the casts...
